File: nim/deathrolls.nim
import std/[strformat, random]

randomize()

var numberOfPlayers = 0;

type Player = ref object
  name: string
  lastRoll: int

proc newPlayer(): Player =
  inc numberOfPlayers
  Player(name: &"Player {$numberOfPlayers}", lastRoll: 1000)
  
method roll(self: Player, startingNumber: int): int =
  self.lastRoll = rand(1..startingNumber)
  echo &"Random! {self.name} rolls a {$self.lastRoll} (out of {$startingNumber})."
  self.lastRoll

var
  currentPlayer = newPlayer()
  otherPlayer = newPlayer()

while otherPlayer.lastRoll != 1:
  discard currentPlayer.roll(otherPlayer.lastRoll)
  swap(currentPlayer, otherPlayer)

echo &"Game over! {currentPlayer.name} wins!"

File: swift/death-rolls.swift
struct Player {
    static var numberOfPlayers: Int = 0

    let name = "Player \(numberOfPlayers + 1)"
    var lastRoll = 1000

    init() {
        Player.numberOfPlayers += 1
    }
    
    mutating func deathRoll(_ startingNumber: Int) {
        lastRoll = Int.random(in: 1...startingNumber)
        print("Random! \(name) rolls a \(lastRoll) (out of \(startingNumber)).")
    }
}

var currentPlayer = Player(), otherPlayer = Player()
while otherPlayer.lastRoll != 1 {
    currentPlayer.deathRoll(otherPlayer.lastRoll)
    swap(&currentPlayer, &otherPlayer)
}

print("Game over! \(currentPlayer.name) wins!")
